src/core_tests/activity_tasks.rs

```diff
@@ -511,12 +552,66 @@ async fn only_returns_cancels_for_desired_queue() {
         }
     };
     assert_matches!(
-        q1_res,
+        &q1_res,
         ActivityTask {
             variant: Some(activity_task::Variant::Cancel(_)),
             ..
         }
     );
+    // act needs to complete to finalize shutdown
+    core.complete_activity_task(ActivityTaskCompletion {
+        task_token: q1_res.task_token,
+        task_queue: "q1".to_string(),
+        result: Some(ActivityExecutionResult::cancel_from_details(None)),
+    })
+    .await
+    .unwrap();
 
     core.shutdown().await;
 }
+
+#[tokio::test]
+async fn can_heartbeat_acts_during_shutdown() {
+    let mut mock_gateway = MockServerGatewayApis::new();
+    mock_gateway
+        .expect_record_activity_heartbeat()
+        .times(1)
+        .returning(|_, _| {
+            Ok(RecordActivityTaskHeartbeatResponse {
+                cancel_requested: false,
+            })
+        });
+    mock_gateway
+        .expect_complete_activity_task()
+        .times(1)
+        .returning(|_, _| Ok(RespondActivityTaskCompletedResponse::default()));
+
+    let core = mock_core(MocksHolder::from_gateway_with_responses(
+        mock_gateway,
+        [],
+        [PollActivityTaskQueueResponse {
+            task_token: vec![1],
+            activity_id: "act1".to_string(),
+            heartbeat_timeout: Some(Duration::from_millis(1).into()),
+            ..Default::default()
+        }],
+    ));
+
+    let act = core.poll_activity_task(TEST_Q).await.unwrap();
+    // Start shutdown before completing the activity
+    let shutdown_fut = core.shutdown();
+    join!(shutdown_fut, async {
```

Comment:
It would be nice to have the test prove that completing the activity is what unblocks core completion